### ISSUE TITLE
CPU persistent compilation cache: cached executables are always rejected on reload (cpu_aot_loader treats LLVM tuning attrs +prefer-no-gather/+prefer-no-scatter as required host CPU features)

### ISSUE THREAD

**Issue by enricobarausse** [jax-ml/jax#38445]
### Description

Description
With the persistent compilation cache enabled on the CPU backend, executables written to the cache by one process are rejected on reload by a subsequent process on the same host (fresh cache directory, identical hardware). Every cache entry fails the XLA:CPU AOT loader's machine-compatibility check with:

Loading XLA:CPU AOT result. Target machine feature +prefer-no-gather is not
supported on the host machine. Machine type used for XLA:CPU compilation
doesn't match the machine type for execution.
(and the same for +prefer-no-scatter). The result is that the CPU persistent cache never hits: every process recompiles everything, silently except for the E-level log lines.

Environment
jax / jaxlib: reproduced identically on 0.9.1, 0.9.2, and 0.10.1 (latest; error at cpu_aot_loader.cc:195 on 0.10.1, :220 on 0.9.x)
Platform: CPU only (JAX_PLATFORMS=cpu)
OS: Linux x86_64 (Rocky Linux 8.10, kernel 4.18)
CPU: Intel Xeon Gold 6238R (Cascade Lake, AVX-512: avx512f/bw/cd/dq/vl/vnni)
Install: plain pip install jax in a clean venv
Reproduction
# repro.py
import jax, jax.numpy as jnp
jax.config.update("jax_enable_x64", True)
jax.config.update("jax_compilation_cache_dir", "/tmp/jax_cpu_cache_repro")
jax.config.update("jax_persistent_cache_min_entry_size_bytes", 0)
jax.config.update("jax_persistent_cache_min_compile_time_secs", 0)

@jax.jit
def f(x):
    return (jnp.sin(x) @ jnp.cos(x.T)).sum()

print(float(f(jnp.ones((512, 128)))))
Run twice in separate processes on the same machine:

JAX_PLATFORMS=cpu python repro.py   # populates the cache, no errors
JAX_PLATFORMS=cpu python repro.py   # prints the loader errors below
Observed
The second run emits one pair of errors per cache entry (6 total here):

E0612 ... cpu_aot_loader.cc:195] Loading XLA:CPU AOT result. Target machine
feature +prefer-no-gather is not  supported on the host machine. Machine type
used for XLA:CPU compilation doesn't match the machine type for execution.
Compile machine features: [+64bit,+adx,+aes,+avx,+avx2,+avx512bw,+avx512cd,
+avx512dq,+avx512f,+avx512vl,+avx512vnni,...,+popcnt,+prefer-no-gather,
+prefer-no-scatter,+prfchw,...,-zu]
vs host machine features: [prfchw,avx,aes,...,avx512f,avx512vnni,...,avx512dq].
This could lead to execution errors such as SIGILL.
(an identical message follows for +prefer-no-scatter; full untrimmed lists available on request). The compile-time feature list always contains +prefer-no-gather,+prefer-no-scatter; the host feature list (CPUID-derived) never does, so the check fails unconditionally and the cached executable is discarded.

Expected
A cache entry compiled on a host should load on that same host; the persistent cache should hit on the second run with no errors.

Analysis
prefer-no-gather and prefer-no-scatter are LLVM x86 tuning attributes (hints to avoid hardware gather/scatter instructions, applied to some Intel targets), not CPUID-backed ISA features. They show up in the compiled target machine's feature string, but they can never appear in the host's detected CPU feature list.

The compatibility check in [xla/backends/cpu/runtime/cpu_aot_loader.cc](https://github.com/enricobarausse/bach_jax/blob/main/xla/backends/cpu/runtime/cpu_aot_loader.cc) (error site at line 195 in current jaxlib 0.10.1, line 220 in 0.9.x) appears to treat every +feature in the compiled target machine as a required host CPU feature, so any executable compiled with these tuning attributes is unconditionally rejected — XLA refuses to load its own output on the machine that produced it.

The check itself also matters beyond the false positive: a real mismatch genuinely risks SIGILL, so simply suppressing the error would be wrong. The fix presumably needs to exclude tuning attributes (or compare only CPUID-meaningful features) in the loader's compatibility check.

Impact
The persistent compilation cache is effectively non-functional for the CPU backend on affected x86 hosts (every entry written is rejected on read; every process pays full compile cost).
The same false-positive class could mask, or be conflated with, genuine ISA mismatches that do risk SIGILL

### System info (python version, jaxlib version, accelerator, etc.)

jax / jaxlib: reproduced identically on 0.9.1, 0.9.2, and 0.10.1 (latest; error at cpu_aot_loader.cc:195 on 0.10.1, :220 on 0.9.x)
Platform: CPU only (JAX_PLATFORMS=cpu)
OS: Linux x86_64 (Rocky Linux 8.10, kernel 4.18)
CPU: Intel Xeon Gold 6238R (Cascade Lake, AVX-512: avx512f/bw/cd/dq/vl/vnni)
Install: plain pip install jax in a clean venv

**Comment by notandruu** [comment]
Triage note: the rejection comes from XLA's CPU AOT loader compatibility check (xla/backends/cpu/runtime/cpu_aot_loader.cc in openxla/xla), which requires every feature recorded at compile time to be present on the host. `+prefer-no-gather` and `+prefer-no-scatter` are LLVM x86 *tuning* preferences (set for some microarchitectures to avoid slow gather/scatter), not ISA features, so the host CPUID-derived feature list never contains them and the check can never pass on machines where the compiler sets them, including the exact machine that produced the cache entry. A fix would likely filter tuning-only flags (or anything with a `prefer-` prefix) out of the compatibility comparison on the XLA side. Filing this pointer here since the fix belongs in openxla/xla rather than this repo.
